### ISSUE TITLE
Result of ring-handler doesn't conform to spec

### ISSUE THREAD

**Issue by conan** [metosin/reitit#83]
It seems that the thing that is returned by `reitit.ring/ring-handler` doesn't conform to the [:ring/handler](https://github.com/ring-clojure/ring-spec/blob/master/src/ring/core/spec.clj#L178) spec:

```
(s/explain :ring/handler (rr/ring-handler (rr/router ["" #(prn "")])))
In: [:args 0] val: :sync fails spec: :ring/request at: [:sync+async :fn :sync :args :request] predicate: map?
In: [:args 0] val: :sync fails spec: :ring/request at: [:sync+async :fn :async :args :request] predicate: map?
In: [:ret] val: [:async nil] fails spec: :ring/response at: [:sync+async :fn :sync :ret] predicate: map?
val: nil fails spec: :ring/response at: [:sync :ret] predicate: map?
val: ({:server-port 1, :server-name "", :remote-addr "", :uri "/", :scheme :http, :protocol "HTTP/1.1", :headers {}, :request-method :a} #object[clojure.spec.alpha$fspec_impl$reify__2451$fn__2454 0x658e90f6 "clojure.spec.alpha$fspec_impl$reify__2451$fn__2454@658e90f6"] #object[clojure.spec.alpha$fspec_impl$reify__2451$fn__2454 0x343045e0 "clojure.spec.alpha$fspec_impl$reify__2451$fn__2454@343045e0"]) fails spec: :ring.async/handler at: [:async] predicate: (apply fn),  Assert failed: In: [0] val: nil fails spec: :ring/response at: [:response] predicate: map?

(pvalid? argspec args)
```
This is a simple example to show the issue, I also see the problem with a more complex (and working) ring handler.

**Comment by ikitommi** [comment]
Hmm... the code for the ring-hander is [here](https://github.com/metosin/reitit/blob/master/modules/reitit-ring/src/reitit/ring.cljc#L119-L154). I'm not sure how to make that to a predefined `fdef`. Ideas welcome.

**Comment by conan** [comment]
It's worth mentioning that in my application, I am setting the default handler using `(rr/create-default-handler)`, and get the error above when trying to validate my handler-creating function using `:ring/handler`

**Comment by miikka** [comment]
That's interesting and there indeed seems to be a problem: `(rr/create-default-handler)` responses do not include `:headers`, which is required by [ring SPEC][SPEC].

[SPEC]: https://github.com/ring-clojure/ring/blob/master/SPEC

```
(s/explain :ring/handler (rr/create-default-handler))
In: [:args 0] val: :sync fails spec: :ring/request at: [:sync+async :fn :sync :args :request] predicate: map?
In: [:args 0] val: :sync fails spec: :ring/request at: [:sync+async :fn :async :args :request] predicate: map?
In: [:ret] val: [:async {:status 404, :body ""}] fails spec: :ring/response at: [:sync+async :fn :sync :ret] predicate: map?
val: {:status 404, :body ""} fails spec: :ring/response at: [:sync :ret] predicate: (contains? % :headers)
val: ({:server-port 1, :server-name "", :remote-addr "", :uri "/", :scheme :http, :protocol "HTTP/1.1", :headers {}, :request-method :a} #object[clojure.spec.alpha$fspec_impl$reify__2451$fn__2454 0x74e6ff3d "clojure.spec.alpha$fspec_impl$reify__2451$fn__2454@74e6ff3d"] #object[clojure.spec.alpha$fspec_impl$reify__2451$fn__2454 0x54b217a0 "clojure.spec.alpha$fspec_impl$reify__2451$fn__2454@54b217a0"]) fails spec: :ring.async/handler at: [:async] predicate: (apply fn),  Assert failed: In: [0] val: {:status 404, :body ""} fails spec: :ring/response at: [:response] predicate: (contains? % :headers)
```

If that is fixed, I can't get spec to complain anymore:

```
(s/explain :ring/handler (rr/ring-handler (rr/router ["" #(prn "")]) (rr/create-default-handler {:not-found (constantly {:status 404, :body "", :headers {}})})))
Success!
```

**Comment by ikitommi** [comment]
Oh. I think they should be optional. Wrote an issue to Ring about it.

**Comment by ikitommi** [comment]
... but as it's only initialized only once, it can be fixed here. PR welcome!

**Comment by ikitommi** [comment]
Not likely to be fixed soon on Ring-side, so PR welcome for this

**Comment by ikitommi** [comment]
Will be fixes in `0.1.2`.
